File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        result = -10**6
        for i in range(len(nums)-2):
            if nums[i]*3 > target + abs(target-result):
                return result
            j = i+1
            k = len(nums)-1
            while j!=k:
                res = nums[i]+nums[j]+nums[k]
                gap = abs(target-res)
                if gap == 0:
                    return res
                if abs(target-result) > gap:
                    result = res
                if res <= target:
                    j += 1
                else:
                    k -= 1
        return result
        

File: 0016-3sum-closest/0016-3sum-closest.c
int comp(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}
int threeSumClosest(int* nums, int numsSize, int target) {
    qsort(nums, numsSize, sizeof(int), comp);
    int result;
    if (target>=0){
        result = INT_MAX;
    }
    else {
        result = INT_MIN;
    }
    int currentres;
    int j;
    int k;
    for (int i=0;i<numsSize-2;i++){
        if (3*nums[i] > target + abs(target-result)) return result;
        j = i+1;
        k = numsSize-1;
        while (j!=k){
            currentres = nums[i]+nums[j]+nums[k];
            if (abs(target-result) > abs(target-currentres)){
                result = currentres;
                if (target == result) return result;
            }
            if (currentres <= target) j+=1;
            else k-=1;
        }
    }
    return result;
}